File: java/word-count/src/main/java/WordCount.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class WordCount {

    Map<String, Integer> phrase(String thePhrase) {

        String[] words = cleanPhrase(thePhrase).split( "\\s+" );

        Histogram<String> h = new Histogram<String>(words);
        return h.histogram;
    }

    private String cleanPhrase(String phrase) {
        return phrase.toLowerCase().replaceAll("[^\\w]", " ");
    }

    public static class Histogram<T> {

        private Map<T, Integer> histogram;

        public Histogram( T[] items ) {
            this.histogram = new HashMap<T, Integer>();

            for ( T item : items ) {
                histogram.put(item, histogram.getOrDefault(item, 0) + 1);
            }
        }

    }
}

File: javascript/word-count/word-count.js
Words = (function() {

  INTL_PUNCT_REGEX = /[\~\?\.\^\[\/\"_\(\)&*\$￥\^\+=`~<>\{\}\[\]\|\-!#%\,\:;@¡§«¶·»¿;·]+/g

  function Words() {
  }

  function histogram(items) {
    var result = {};
    var PROPS_DESCRIPTOR = { value: 0, writable: true, enumerable: true};
    items.forEach(function(item) {
      if (!result.hasOwnProperty(item)) {
        Object.defineProperty(result, item, PROPS_DESCRIPTOR);
      }
      result[item] = result[item] + 1;
    });
    return result;
  }

  function extractWords(phrase) {
    var words = []
    cleanPhrase(phrase).split(/\s+/).forEach(function(word) {
      var cleaned = cleanWord(word);
      if (cleaned) {
        words.push(cleaned);
      }
    });
    return words;
  }

  function cleanPhrase(phrase) {
    return phrase.replace(INTL_PUNCT_REGEX, ' ').toLowerCase()
  }

  function cleanWord(word) {
    return word.replace(/^["'](.*)["']$/, "$1").trim()
  }

  Words.prototype.count = function( phrase ) {
    words = extractWords(phrase);
    return histogram(words);
  };
  return Words;

})();

module.exports = Words;
